pkg/kubelet/cm/cpumanager/topology_hints.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpumanager
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/klog"
+	"math"
+
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/topology"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/socketmask"
+)
+
+func (m *manager) GetTopologyHints(pod v1.Pod, container v1.Container) []topologymanager.TopologyHint {
+	var cpuHints []topologymanager.TopologyHint
+	for resourceObj, amountObj := range container.Resources.Requests {
+		resource := string(resourceObj)
+		requested := int(amountObj.Value())
+		if resource != "cpu" {
+			continue
+		}
+
+		klog.Infof("[cpumanager] Guaranteed CPUs detected: %v", requested)
+
+		topo, err := topology.Discover(m.machineInfo)
+		if err != nil {
+			klog.Infof("[cpu manager] error discovering topology")
+		}
+		var assignableCPUs cpuset.CPUSet
+		containerID, _ := findContainerIDByName(&pod.Status, container.Name)
+		if cset, ok := m.state.GetCPUSet(containerID); ok {
+			klog.Infof("[cpumanager] Reusing pre-assigned CPUSet: %v", cset)
+			assignableCPUs = cset
+		} else {
+			// Otherwise, calculate the assignable CPUs from the topology.
+			assignableCPUs = m.getAssignableCPUs(topo)
+		}
+		klog.Infof("AssignableCPUs: %v", assignableCPUs)
+		cpuAccum := newCPUAccumulator(topo, assignableCPUs, requested)
+
+		//Can we always assume NumSockets and Socket Numbers are the same?
+		socketCount := topo.NumSockets
+		klog.Infof("[cpumanager] Number of sockets on machine (available and unavailable): %v", socketCount)
+		cpuHintsTemp := getCPUMask(socketCount, cpuAccum, requested)
+		CPUsPerSocket := topo.CPUsPerSocket()
+		cpuHints = getPreferred(cpuHintsTemp, requested, CPUsPerSocket)
```

Comment:
We don't need to address this now because the current algorithm seems to do the right thing, but we should consider cleaning this logic up a bit.  It seems like the general flow should be:

1) Get a count of how many guaranteed CPUs have been requested: `requested int`
2) Get a count of how many CPUs are available on each socket: `available []int`
3) Build hints from all possible socket combinations where the request can be granted: `hints []TopologyHint`
4) Set hints with the lowest number of sockets set in their socket mask to `Preferred: true`
5) Set all other hints to `Preferred: false`

If we could distill the code in this function down to just 5 function calls that do these 5 things with these comments above them, that would help with understanding the flow quite a bit.